docs/testing-modes.md

```diff
@@ -1,275 +1,260 @@
 
-1. Specify the initial state of the `Stack` by calling the constructor.
+Let's dissect the structure of the stress test for our `Counter`, here are the steps:
 
-2. We are testing concurrent behaviour of `push(value)`, `pop()` and `size()` operations, so list them all and mark with `@Operation` annotation.
+1. Create an instance of the `Counter` as an initial state.
+2. List operations defined on the `Counter` and mark them with `@Operation` annotation.
+3. Specify the testing mode via the corresponding `Options` class:
+   - for _stress testing mode_ use stress testing options: `StressOptions()`
+4. Run the analysis by invoking the `StressOptions().check(..)` on the testing class.
 
-3. Choose `StressOptions()` to run the test in the stress mode.
-
-4. Run the test by invoking the `StressOptions().check(..)` function on the testing class.
+Below is the full code of the test:
 
 ```kotlin
-class StackTest {
-    private val s = Stack<Int>()
+import org.jetbrains.kotlinx.lincheck.annotations.Operation
+import org.jetbrains.kotlinx.lincheck.check
+import org.jetbrains.kotlinx.lincheck.strategy.stress.StressOptions
+import org.jetbrains.kotlinx.lincheck.verifier.VerifierState
+import org.junit.Test
+
+class CounterTest {
+   private val c = Counter() // initial state
 
-    @Operation fun push(value: Int) = s.push(value)
-    @Operation fun pop() = s.pop()
-    @Operation fun size() = s.size
+   // operations on the Counter
+   @Operation fun inc() = c.inc()
+   @Operation fun get() = c.get()
 
-    @Test // stress testing mode
-    fun stressTest() = StressOptions().check(this::class.java)
+   @Test // run the test
+   fun runStressTest() = StressOptions().check(this::class)
 }
 ```
 
-> Get the full code [here](../src/jvm/test/org/jetbrains/kotlinx/lincheck/test/guide/StackTest.kt).
- 
-Run the `stressTest()` function. It will fail with the following output:
+### How stress mode works under the hood
 
-```text
-= Invalid execution results =
-Parallel part:
-| push(10): void | pop():  10       |
-|                | size(): -1 [-,1] |
-```
+At first, some parallel execution scenario is generated from the operations that were marked with `@Operation` annotation.
 
-The stress test revealed the incorrect behaviour of `size()` implementation: negative return value. 
-Here we come again to the main advantage of the model checking mode: providing a trace to reproduce the error.
+Then `Lincheck` starts real threads, synchronizes them in an active manner to guarantee that the operations are performed in parallel, 
+and executes the operations, repeating the run many times in the hope to hit an interleaving that produces an incorrect result. 
+
+To increase the number of different interleavings, we also add random busy-wait cycles between the operations and inside them. 
+
+Below is the high-level picture of how one of the stress executions for the `Counter` may be performed.
+
+![Stress execution of the Counter](../docs/images/counter_stress_mode.png)
```

Comment:
I think the picture of stress testing execution may be quite useful for getting the idea.

Though we should do something about it's size